File: 2024/Day01.scala
object Day01:
  @main def main: Unit =
    val input = parse()
    println(part1(input))
    println(part2(input))

  def parse(): (Seq[Int], Seq[Int]) =
    val lines = io.Source.fromFile("day01.txt").getLines.toSeq
    val linesSeq: Seq[(Int, Int)] = lines.map:
      case s"$a   $b" => (a.toInt, b.toInt)

    linesSeq.unzip

  def part1(lists: (Seq[Int], Seq[Int])): Int =
    val left = lists(0).sorted
    val right = lists(1).sorted

    left.zip(right).map((l, r) => math.abs(l - r)).sum

  def part2(lists: (Seq[Int], Seq[Int])): Int =
    val left = lists(0)
    val frequencies = lists(1).groupBy(identity).view.mapValues(_.size)

    left.fold(0): (acc, i) =>
      acc + (i * frequencies.getOrElse(i, 0))

File: 2024/day01.exs
{left, right} =
  File.stream!("day01.txt")
  |> Stream.map(fn str ->
    String.split(str) |> Enum.map(&String.to_integer/1) |> List.to_tuple()
  end)
  |> Enum.unzip()

Enum.zip_reduce(Enum.sort(left), Enum.sort(right), 0, fn l, r, acc -> acc + abs(l - r) end)
|> IO.inspect()

frequencies = Enum.frequencies(right)

Enum.reduce(left, 0, fn i, acc -> acc + i * Map.get(frequencies, i, 0) end) |> IO.inspect()
